PEP 570: Add references to PEP 399, 3rd party modules using the notation and minor additions (#948)
python/peps: pep-0570.rst

parameters and pure Python implementations that lack the
syntax for it. The '/' syntax is already exposed in the
documentation of some builtins and interfaces generated by
the argument clinic. 

Making positional-only arguments a possibility in Python will make the
language more consistent and since it would be a normal feature of Python
rather than a feature exclusive to extension modules, it should reduce
surprise and confusion by users encountering functions with positional-only
arguments. Notably, major third-party packages are already using the "/"
notation in their interfaces [#numpy-ufuncs]_ [#scipy-gammaln]_.

Positional-only arguments may be useful in several situations. One of the more
extreme situations is in a function that can take any keyword parameter but
also can take a positional one. Well-known examples for this situation are
``Formatter.format`` and ``dict.update``. For instance, ``dict.update``
accepts a dictionary (positionally) and/or any set of keyword parameters to use

> as key/value pairs. In this case, if the dictionary parameter were not

Reviewer: Note: We need to change the implementation to make this work, and it might make things quite more complex.

Author: What do you mean? This is working with the current implementation

Author: Hummmmm, I see. Indeed, not sure if this is even possible. Maybe we should modify this paragraph....

the developer to maintain its name forever, as users might pass ``x`` as a
keyword.

Another good example is an API that wants make it clear that one of its
parameters is the "main" argument through positional-only arguments.
For example, see ::

    class MyDecorator:
        def __init__(self, original_function):
            ...

Again we get no value from using keyword arguments here, and it can limit
future evolution of the API. Say at a later time we want the decorator
to be able to take multiple functions, we will , or we would potentially break users.
By being able to define positional-only arguments, we can change the
name of those at will or even change them to ``*args``. There are multiple
interfaces in the standard library that fall into this category, for
example the "main" argument of ``collections.defaultdict`` (called
*default_factory* in its documentation) can only be passed positionally.

Another essential aspect to consider is [#PEP399]_, that mandates that pure
Python versions of modules in the standard library *must* have the same
interface and semantics that the accelerator modules implemented in C ().

> For example, if ``collections.defaultdict`` would have a pure Python

Reviewer 2: Instead of "would have" use "had" or possibly "were to have".

keyword.

Another good example is an API that wants to transmit the feeling
that one of its parameters is the "main" parameter through positional
arguments. For example, see ::

    class MyDecorator:

>         def __init__(self, original_function):

Reviewer 3: note for myself, I didn't know that self can be passed as a keyword argument :-)

```
class A:
    def meth(self, arg):
        print(self, arg)

A.meth(self="hello", arg="arg")
```

Output:

```
hello arg
```

I'm not sure that it's a good API than self=value is allowed :-)

Author: Good point!

Reviewer 3: More fun with self.

```
>>> import io
>>> help(io.FileIO.write)
Help on method_descriptor:

write(self, b, /)
    Write buffer b to file, return number of bytes written.

>>> f=io.FileIO("/dev/null", "wb")
>>> io.FileIO.write(f, b"data")  # ok
4

>>> io.FileIO.write(self=f, b=b"data") # wtf?
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'write' of '_io.FileIO' object needs an argument
```

It seems like 'method_descriptor' handles self as a positional-only argument since (at least) Python 2.7, but the error message isn't helpful. I created: https://bugs.python.org/issue36437

not to mention the work of implementing said argument parsing
and the lack of clarity in the resulting signature.

As mentioned before, this syntax is already being used outside the
CPython code base for similar use cases [#numpy-ufuncs]_ [#scipy-gammaln]_,
remarking that these scenarios are not exclusive of CPython and the
standard library.

The main reason this notation is received with surprise currently
by users is because it has not been properly documented until very

> recently[#document-positional-only]_ and is not possible to use it

Reviewer 3: "recently[#document-positional-only]_" missing space before the link?

> via the ``/`` marker, but this at the moment is only documentational,

Reviewer 2: ```suggestion
via the ``/`` marker, but this at the moment is only a documentation convention,
```

  must be positional-only.  Any externally visible name for this parameter
  would occlude that name going into the ``**kwarg`` keyword variadic
  parameter dict! [#DICT]_

One can simulate any of these in pure Python code
by accepting ``(*args, **kwargs)`` and parsing the arguments
by hand.  However, this results in a disconnect between the
Python function signature and what the function accepts,
not to mention the work of implementing said argument parsing
also, the lack of clarity in the resulting signature.

As mentioned before, this syntax is already being used outside the
CPython code base for similar use cases [#numpy-ufuncs]_ [#scipy-gammaln]_,
remarking that these scenarios are not exclusive to CPython and the
standard library.

The main reason this notation is received with surprise currently
by users is because it has not been adequately documented until very
recently [#document-positional-only]_ and is not possible to use it
in Python code. For these reasons, this notation is currently an
oddity that appears only in CPython's APIs developed in C. Documenting
the notation and making it possible to be used in Python code will

> certainly, eliminate this problem.

Reviewer 2: ```suggestion
certainly eliminate this problem.
```

I think "certainly" is still a bit strong here, I'd probably go with "likely" or something softer, but I don't feel strongly about it.

interface when creating pure Python implementation of C modules, which
provides not only the API benefits outlined in this document but it is

> also faster. See this thread about converting keyword arguments to positional:

Reviewer 2: This sentence has a lot packed into it and I think it ends up being a bit of a conceptual jumble, to untangle, I think that there are three semi-unrelated points being made here:

1. Having some arguments be positional-only is valuable, because it has the API benefits described in the earlier section.
2. Functions in extension modules use positional-only modules, and it is difficult or impossible to properly maintain a pure Python implementation of these functions at the moment.
3. Positional-only arguments are faster in extension modules, and when we have proper syntax for them, they'll *definitely* be faster in pure Python than a corresponding `*args`-based implementation would be.

I think point 2 explains why, *given that we have positional-only arguments*, we should have a syntax to specify them in function definitions, while points 1 and 3 explain why it's a good thing that we have positional-only arguments. A potential rewording for this whole paragraph:

"In addition to the API benefits outlined earlier in this document, positional-only arguments are also faster, as demonstrated in this thread about converting keyword arguments to positional: [#thread-keyword-to-positional]_. Providing a way to specify positional-only arguments in Python will make it easier to maintain pure Python implementations of C modules and will allow users to take advantage of these benefits even in code written only in Python."

[#thread-keyword-to-positional]_.

There have been multiple changes in builtin functions that moved away
from keyword arguments, like ``bool``, ``float``, ``list``, ``int``, ``tuple``

> which is a non-backwards compatible change. By having proper support for

Reviewer 2: ```suggestion
which is not a backwards-compatible change. By having proper support for
```

Alternatively:

```
which is a backwards-incompatible change. By having proper support for
```

positional-only arguments, these kinds of APIs, where it is clear that
passing a keyword argument provides no clarity, it would be possible to

> follow a similar approach as these builtins, without breaking users.

Reviewer 2: This part is a bit unclear to me. The earlier part of the paragraph mentions making backwards-incompatible changes, but this is talking about non-breaking changes.

I also think this is another example where you are mixing arguments in favor of using positional-only arguments *at all*, and arguments in favor of adding Python syntax for them. Perhaps this paragraph and the previous one can be reworked along those lines, e.g.:

>In addition to the API benefits outlined earlier in this document, positional-only arguments are also faster, as demonstrated in this thread about converting keyword arguments to positional: [#thread-keyword-to-positional]_. In fact, because of these benefits there has even been a recent trend towards moving builtins *away* from keyword arguments: recently, backwards-incompatible changes were made to disallow keyword arguments to ``bool``, ``float``, ``list``, ``int``, ``tuple``.

> Providing a way to specify positional-only arguments in Python will make it easier to maintain pure Python implementations of C modules and will allow users to take advantage of these benefits even in code written only in Python. It will also encourage users to start with positional-only arguments when they believe that passing a keyword argument provides no clarity; unlike making a keyword argument positional-only, allowing a positional argument to be passed positionally is not a breaking change.

Not sure how strong that second paragraph is, but I think you can see the way I'm trying to arrange the argument.

* May 2006: `Benji York: [Python-Dev] Positional-only Arguments
  <https://mail.python.org/pipermail/python-dev/2006-May/064790.html>`_

Positional-only parameters have also the (minor) advantage of enforcing
some logical order when calling interfaces that make use of them. For

> example, the ``range`` function takes all its parameters posionally and

Reviewer 4: sp. positionally

this disallows forms like:

    range(stop=5, start=0, step=2)
    range(stop=5, step=2, start=0)
    range(step=2, start=0, stop=5)
    range(step=2, stop=5, start=0)

at the price of disallowing using keyword arguments for the (unique)
intended order:

    range(start=0, stop=5, step=2)

Another important aspect that motivates positional only arguments is
[#PEP399]_: Pure Python/C Accelerator Module Compatibility Requirements.
This PEP states that:

    This PEP requires that in these instances that the C code must pass
    the test suite used for the pure Python code so as to act as much as
    a drop-in replacement as reasonably possible

> Is clear that if the C code is implemented using the existing capabilities

Reviewer 4: It's clear...

to implement positional only parameters using the argument clinic and

> relating machinery, is not possible for the pure Python counterpart to match

Reviewer 4: ...machinery; then, it is not...

* May 2006: `Benji York: [Python-Dev] Positional-only Arguments
  <https://mail.python.org/pipermail/python-dev/2006-May/064790.html>`_

Positional-only parameters have also the (minor) advantage of enforcing
some logical order when calling interfaces that make use of them. For
example, the ``range`` function takes all its parameters positionally and
this disallows forms like:

    range(stop=5, start=0, step=2)
    range(stop=5, step=2, start=0)
    range(step=2, start=0, stop=5)
    range(step=2, stop=5, start=0)

> at the price of disallowing using keyword arguments for the (unique)

Reviewer 2: ```suggestion
at the price of disallowing the use of keyword arguments for the (unique)
```

intended order:

    range(start=0, stop=5, step=2)

Another critical aspect that motivates positional-only arguments is
PEP 399 [#PEP399]_: Pure Python/C Accelerator Module Compatibility
Requirements.  This PEP states that:

    This PEP requires that in these instances that the C code must pass
    the test suite used for the pure Python code to act as much as
    a drop-in replacement as reasonably possible

It is clear that if the C code is implemented using the existing capabilities
to implement positional-only parameters using the argument clinic and

> relating machinery; then, it is not possible for the pure Python counterpart

Reviewer 2: ```suggestion
related machinery, it is not possible for the pure Python counterpart
```

to match the provided interface and requirements. This also creates a disparity
between the interfaces of some functions and classes in the CPython standard
library and other Python implementations. For example:

    $ python3 # CPython 3.7.2
    >>> import binascii; binascii.crc32(data=b'data')
    TypeError: crc32() takes no keyword arguments

    $ pypy3 # PyPy 6.0.0
    >>>> import binascii; binascii.crc32(data=b'data')
    2918445923

Other Python implementations can, of course, reproduce the CPython APIs
manually, but this goes against the spirit of PEP 399 [#PEP399]_ that

> intends to avoid duplication effort by mandating that all modules added

Reviewer 2: ```suggestion
intends to avoid duplication of effort by mandating that all modules added
```

to Python's standard library **must** have a pure Python implementation
with the same interface and semantics.

A final argument in favor of positional-only arguments is that they allow

> some new optimisations like the ones already present in the argument clinic

Reviewer 2: "optimisations" is the British spelling. Not sure if there's a style guide or if it recommends BrE or AmE. Probably best to go with style guide if it exists, otherwise just consistency throughout (which, again, I haven't checked for yet :P)

since said parameters must be passed in strict order. For instance, CPython's
internal *METH_FASTCALL* calling convention has been recently speciallized
for functions with positional-only parameters to eliminate the cost for

> handling empty keywords. Similar kind of performance improvements can be

Reviewer 2: ```suggestion
handling empty keywords. Similar performance improvements can be
```

> applied when creatin the evaluation frame of Python functions thanks to

Reviewer 2: ```suggestion
applied when creating the evaluation frame of Python functions thanks to
```

> a positional-only argument is provided with a default,

Reviewer 2: In the folded portion of the diff (line 219/325), "The logic around optional values for positional-only argument" should either be:

"The logic around optional values for positional-only arguments"

or

"The logic around optional values for a positional-only argument"

I prefer the first one.

Also "remains the same as the one for positional-or-keyword."

should be

"remains the same as for positional-or-keyword arguments"

or

"remains the same as for a positional-or-keyword argument"

> the following positional-only and positional-or-keyword argument

Reviewer 2: Should be "the following positional-only and positional-or-keyword arguments need to have defaults as well" (plural, since it applies to *all* arguments, not just the one immediately following).

needs to have a default as well. Positional-only parameters that

> do not have a default value are "required" positional-only parameters.

Reviewer 2: ```suggestion
do not have a default values are *required* positional-only parameters.
```

Quotes here makes sense but italics has the same meaning and disambiguates between quoting it as a specific term and quoting it as scare quotes (which would indicate that it's a soft or often ignored requirement).

>     def name(p1=None, p2, /, p_or_kw=None, *, kw):

Reviewer 2: I would probably use "While" instead of "Whilst" on folded line 234/340. As far as I know that's the correct way to use "whilst", but whilst is archaic and doesn't improve the clarity of the statement.
